lightning/src/ln/channelmanager.rs

```diff
@@ -71,6 +71,181 @@ use core::time::Duration;
+		hmac.input(&metadata_bytes);
+		hmac.input(&payment_hash.0);
+		let hmac_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let mut iv_bytes = [0 as u8; IV_LEN];
+		iv_bytes.copy_from_slice(&hmac_bytes[..IV_LEN]);
+
+		Ok(construct_payment_secret(&iv_bytes, &metadata_bytes, &keys.metadata_key))
+	}
+
+	fn construct_metadata_bytes(min_value_msat: Option<u64>, user_pmt_hash: bool, invoice_expiry_delta_secs: u32, highest_seen_timestamp: u64) -> Result<[u8; 16], APIError> {
+		if min_value_msat.is_some() && min_value_msat.unwrap() > MAX_VALUE_MSAT {
+			return Err(APIError::APIMisuseError { err: format!("min_value_msat of {} greater than total 21 million bitcoin supply", min_value_msat.unwrap()) });
+		}
+
+		let mut min_amt_msat_bytes: [u8; 8] = match min_value_msat {
+			Some(amt) => amt.to_be_bytes(),
+			None => [0; 8],
+		};
+		// Set the payment method bits (the top 3 bits of the metadata bytes) to 001
+		// indicating the payment type is paying a user-generated payment hash.
+		if user_pmt_hash {
+			min_amt_msat_bytes[0] |= 1 << 5;
+		}
+
+		// We assume that highest_seen_timestamp is pretty close to the current time - it's updated when
+		// we receive a new block with the maximum time we've seen in a header. It should never be more
+		// than two hours in the future.  Thus, we add two hours here as a buffer to ensure we
+		// absolutely never fail a payment too early.
+		// Note that we assume that received blocks have reasonably up-to-date timestamps.
+		let expiry_bytes = (highest_seen_timestamp + invoice_expiry_delta_secs as u64 + 7200).to_be_bytes();
+
+		let mut metadata_bytes: [u8; METADATA_LEN] = [0; METADATA_LEN];
+		{
+			let (min_amt_msat_slice, expiry_slice) = metadata_bytes.split_at_mut(8);
+			min_amt_msat_slice.copy_from_slice(&min_amt_msat_bytes);
+			expiry_slice.copy_from_slice(&expiry_bytes);
+		}
+
+		Ok(metadata_bytes)
+	}
+
+	fn construct_payment_secret(iv_bytes: &[u8; IV_LEN], metadata_bytes: &[u8], metadata_key: &[u8; METADATA_KEY_LEN]) -> PaymentSecret {
+		let mut payment_secret_bytes: [u8; 32] = [0; 32];
+		{
+			let (iv_slice, encrypted_metadata_slice) = payment_secret_bytes.split_at_mut(16);
+			iv_slice.copy_from_slice(iv_bytes);
+
+			let chacha_block = ChaCha20::get_single_block(metadata_key, iv_bytes);
+			for i in 0..METADATA_LEN {
+				encrypted_metadata_slice[i] = chacha_block[i] ^ metadata_bytes[i];
+			}
+		}
+		PaymentSecret(payment_secret_bytes)
+	}
+
+	// Check that an inbound payment's `payment_data` field is sane.
+	//
+	// LDK does not store any data for pending inbound payments. Instead, we construct our payment
+	// secret (and, if supplied by LDK, our payment preimage) to include encrypted metadata about the
+	// payment.
+	//
+	// The metadata is constructed as:
+	//   payment method (3 bits) || payment amount (8 bytes - 3 bits) || expiry (8 bytes)
+	//
+	// Then on payment receipt, we verify in this method that the payment preimage and payment secret
+	// match what was constructed.
```

Comment:
Could you make this the `inbound_payment` module documentation (`///`) and expand it a bit to describe the payment methods and the encryption scheme?

Reply:
Added more info! Fyi there was an earlier version that got cut down after [this feedback](https://github.com/lightningdevkit/rust-lightning/pull/1177#discussion_r755485730)

Reply:
Hmm... I can't see the full doc snippet it was in response to. Let me know if this contradicts the other feedback.

I guess, as written, it's clear how the functions relate, but it's unclear to the casual reader what a payment method is and how it affects the encryption scheme without trying to parse all the code across both functions. For the latter, it would be nice if the `inbound_payment` module described why there are different payment methods and what is accomplished by using different encryption schemes. Or maybe describe each on the corresponding enum variants? FWIW, this is very much developer documentation, so I think posterity would mostly want to understand the reasoning behind having and using these schemes, not necessarily how they are defined in detail.

Reply:
My comment was more of a "we don't need to bog the user down in specific cryptographic implementation details here, keep the docs high level" not "we shouldn't describe that the thing is encrypted and MAC'd data"

Reply:
Agreed! To clarify, by "developer documentation" I meant internal docs for LDK contributors.